src/annotator/integrations/test/vitalsource-test.js

```diff
@@ -0,0 +1,215 @@
+      fakeGuest.injectClient.resetHistory();
+
+      fakeViewer.loadNextChapter();
+      await delay(0);
+
+      assert.calledWith(fakeGuest.injectClient, fakeViewer.contentFrame);
+    });
+
+    it('does not allow annotation in the container frame', async () => {
+      assert.equal(integration.canAnnotate(), false);
+
+      // Briefly check the results of the stub methods.
+      assert.instanceOf(await integration.anchor(), Range);
+      assert.throws(() => integration.describe());
+      assert.equal(integration.fitSideBySide(), false);
+      assert.deepEqual(await integration.getMetadata(), {
+        title: '',
+        link: [],
+      });
+      assert.equal(await integration.uri(), document.location.href);
+      assert.equal(integration.contentContainer(), document.body);
+      await integration.scrollToAnchor({}); // nb. No assert, this does nothing.
+    });
+  });
+
+  describe('VitalSourceContentIntegration', () => {
+    let integration;
+
+    beforeEach(() => {
+      integration = new VitalSourceContentIntegration();
+    });
+
+    afterEach(() => {
+      integration.destroy();
+    });
+
+    it('allows annotation', () => {
+      assert.equal(integration.canAnnotate(), true);
+    });
+
+    it('does not support side-by-side mode', () => {
+      assert.equal(integration.fitSideBySide(), false);
+    });
+
+    it('stops mouse events from propagating to parent frame', () => {
+      const events = ['mousedown', 'mouseup', 'mouseout'];
+
+      for (let eventName of events) {
+        const listener = sinon.stub();
+        document.addEventListener(eventName, listener);
+
+        const event = new Event(eventName, { bubbles: true });
+        document.body.dispatchEvent(event);
+        assert.notCalled(listener);
+
+        document.removeEventListener(eventName, listener);
+      }
+    });
+
+    it('delegates to HTML integration for anchoring', async () => {
+      await integration.contentContainer();
+      assert.calledWith(fakeHTMLIntegration.contentContainer);
+
+      const range = new Range();
+      await integration.describe(range);
+      assert.calledWith(fakeHTMLIntegration.describe, range);
+
+      const selectors = [{ type: 'TextQuoteSelector', exact: 'foobar' }];
+      await integration.anchor(selectors);
+      assert.calledWith(fakeHTMLIntegration.anchor, selectors);
+
+      const anchor = {};
+      await integration.scrollToAnchor(anchor);
+      assert.calledWith(fakeHTMLIntegration.scrollToAnchor, anchor);
+    });
+
+    describe('#getMetadata', () => {
+      it('returns book metadata', async () => {
+        const metadata = await integration.getMetadata();
+        assert.equal(metadata.title, document.title);
+        assert.deepEqual(metadata.link, []);
+      });
+    });
+
+    describe('#uri', () => {
+      beforeEach(() => {
+        const bookURI =
+          '/books/abc/epub/OPS/xhtml/chapter_001.html?ignoreme#cfi=/foo/bar';
+        history.pushState({}, '', bookURI);
```

Comment:
This is clever. I don't know that it would occur to me to use `history` to do this testing.